src/Pos/DB/Block/Block.hs

```diff
@@ -0,0 +1,302 @@
+
+-- | Load blunds which have depth less than given.
+loadBlundsByDepth
+    :: (SscHelpersClass ssc, MonadDB m)
+    => Word -> HeaderHash -> m (NewestFirst [] (Blund ssc))
+loadBlundsByDepth = loadDataByDepth getBlundThrow (const True)
+
+-- | Load blocks starting from block with header hash equal to given hash
+-- and while @predicate@ is true.
+loadBlocksWhile
+    :: (SscHelpersClass ssc, MonadDB m)
+    => (Block ssc -> Bool) -> HeaderHash -> m (NewestFirst [] (Block ssc))
+loadBlocksWhile = loadDataWhile getBlockThrow
+
+-- | Load headers starting from block with header hash equal to given hash
+-- and while @predicate@ is true.
+loadHeadersWhile
+    :: (SscHelpersClass ssc, MonadDB m)
+    => (BlockHeader ssc -> Bool)
+    -> HeaderHash
+    -> m (NewestFirst [] (BlockHeader ssc))
+loadHeadersWhile = loadDataWhile getHeaderThrow
+
+-- | Load headers which have depth less than given.
+loadHeadersByDepth
+    :: (SscHelpersClass ssc, MonadDB m)
+    => Word -> HeaderHash -> m (NewestFirst [] (BlockHeader ssc))
+loadHeadersByDepth = loadDataByDepth getHeaderThrow (const True)
+
+-- | Load headers which have depth less than given and match some criterion.
+loadHeadersByDepthWhile
+    :: (SscHelpersClass ssc, MonadDB m)
+    => (BlockHeader ssc -> Bool)
+    -> Word
+    -> HeaderHash
+    -> m (NewestFirst [] (BlockHeader ssc))
+loadHeadersByDepthWhile = loadDataByDepth getHeaderThrow
+
+----------------------------------------------------------------------------
+-- Initialization
+----------------------------------------------------------------------------
+
+prepareBlockDB
+    :: forall ssc m.
+       (SscHelpersClass ssc, MonadDB m)
+    => GenesisBlock ssc -> m ()
+prepareBlockDB g = do
+    putBi lastBlundFileKey (FixedSizeInt (0::Int))
+    putBlund (Left g) def
+
+----------------------------------------------------------------------------
+-- Keys
+----------------------------------------------------------------------------
+
+blockIndexKey :: HeaderHash -> ByteString
+blockIndexKey h = "b" <> convert h
+
+blockAuxKey :: HeaderHash -> ByteString
+blockAuxKey h = "a" <> convert h
+
+lastBlundFileKey :: ByteString
+lastBlundFileKey = "last_blund_file"
+
+----------------------------------------------------------------------------
+-- Helpers
+----------------------------------------------------------------------------
+
+getBi
+    :: (MonadDB m, Bi v)
+    => ByteString -> m (Maybe v)
+getBi k = rocksGetBi k =<< getBlockIndexDB
+
+putBi
+    :: (MonadDB m, Bi v)
+    => ByteString -> v -> m ()
+putBi k v = rocksPutBi k v =<< getBlockIndexDB
+
+-- delete :: (MonadDB m) => ByteString -> m ()
+-- delete k = rocksDelete k =<< getBlockIndexDB
+
+getData :: (MonadIO m, MonadCatch m, Bi v)
+        => Word32 -- Offset
+        -> Word32 -- Length
+        -> FilePath
+        -> m v
+getData offset len fp = liftIO $ withFile fp ReadMode $ \h -> do
+    hSeek h AbsoluteSeek (fromIntegral offset)
+    bs <- BS.hGet h (fromIntegral len)
```

Comment:
I don't think it's more efficient. If you use lazy IO and then lazy decode, you'll get streaming for free. If you do everything strictly, you have to load the whole thing into RAM.